extra/migrate_settings.rb

```diff
@@ -1,108 +1,150 @@
+      ::Proxy::Migration.inject_migrations_instance(@migrations = ::Proxy::Migrations.new)
+      migration_files.each { |file| require file }
+    end
+
+    def migrate
+      raise "Migrations dir '#{migrations_dir_path}' doesn't exist" unless Dir.exist?(migrations_dir_path)
+      raise "Settings dir '#{config_src_dir}' doesn't exist" unless Dir.exist?(config_src_dir)
+
+      FileUtils.mkdir_p(result_dir_path)
+      last_migration = File.read(migration_state_file_path) if File.exist?(migration_state_file_path)
+
+      load_migrations(migrations_dir_path)
+
+      executed_migrations = []
+      migrations = @migrations.since_last(last_migration)
+      migrations.each do |migration|
+        m = migration.new(working_dir_path)
+
+        create_migration_dirs(m.src_dir, m.dst_dir)
+        if  migration == migrations.first
+          copy_original_configuration(m.src_dir)
+        else
+          copy_previous_migration_results(executed_migrations.last.dst_dir, m.src_dir)
+        end
 
-def write_to_files(output,unknown)
-  modules.each do |m|
-    next if output[m] == {}
-    File.open("#{m}.yml",'w') do |f|
-      f.write(output[m].to_yaml)
-      if m == :settings && unknown != {}
-        f.write "\n# Unparsed options, please review\n"
-        f.write(unknown.to_yaml.gsub(/^---/,''))
+        m.migrate
+
+        if migration == migrations.last
+          copy_to_results_dir(m.dst_dir, result_dir_path)
+        end
+
+        executed_migrations << m
       end
+
+      persist_migration_state(executed_migrations.empty? ? last_migration : executed_migrations.last.class.name)
+    end
+
+    def create_migration_dirs(src_dir, dst_dir)
+      FileUtils.mkdir_p(src_dir)
+      FileUtils.mkdir_p(File.join(dst_dir, "settings.d"))
+    end
+
+    def copy_original_configuration(src_dir)
+      FileUtils.cp_r(File.join(config_src_dir, '.'), src_dir)
+    end
+
+    def copy_previous_migration_results(result_dir, src_dir)
+      FileUtils.cp_r(File.join(result_dir, '.'), src_dir)
+    end
+
+    def copy_to_results_dir(dst_dir, result_dir)
+      FileUtils.cp_r(File.join(dst_dir, '.'), result_dir)
+    end
+
+    def persist_migration_state(last_executed_migration)
+      migration_state = File.open(File.join(result_dir_path, "migration_state"), "w")
+      migration_state.write(last_executed_migration)
+    ensure
+      migration_state.close unless migration_state.nil?
     end
   end
 end
 
-# When running as a script
 if __FILE__ == $0 then
-  orig_file = ARGV[0] || '/etc/foreman-proxy/settings.yml'
-  data      = YAML.load_file(orig_file)
+  if ARGV.size < 4
+    p "Usage: migrate.rb <config_src_dir> <working_dir_path> <migrations_dir_path> <migrations_state_file_path>"
+  end
 
-  output,unknown = migrate(data.dup)
-  exit(1) if data == output[:settings] # it looks like we already have the new style settings
-  write_to_files(output,unknown)
+  config_src_dir = ARGV[0]
+  working_dir_path = ARGV[1]
```

Comment:
The working dir is designed to hold both temporary (working) files and the results, right?  So the caller would run this script and then copy the results back over the real /etc/ dir?